crates/node/src/lib.rs

```diff
@@ -556,23 +557,35 @@ impl Node {
         payload: Vec<u8>,
         aliases: &[Alias<PublicKey>],
     ) -> Result<Vec<u8>, CallError> {
-        let payload_str = String::from_utf8(payload).map_err(|_| CallError::InternalError)?;
+        if aliases.is_empty() {
+            return Ok(payload);
+        }
+
+        let mut result = Vec::with_capacity(payload.len());
+        let mut remaining = &payload[..];
 
-        let mut modified_payload = payload_str;
         for alias in aliases {
             let placeholder = format!("{{{}}}", alias.as_ref());
-            if modified_payload.contains(&placeholder) {
+            let placeholder_bytes = placeholder.as_bytes();
+
+            while let Some(pos) = memmem::find(remaining, placeholder_bytes) {
+                result.extend_from_slice(&remaining[..pos]);
+
                 let public_key = self
                     .ctx_manager
                     .resolve_alias(*alias, Some(context_id))
-                    .map_err(|_| CallError::InternalError)?
+                    .map_err(|_| CallError::AliasResolutionFailed)?
                     .ok_or(CallError::InternalError)?;
 
-                modified_payload = modified_payload.replace(&placeholder, &public_key.to_string());
+                result.extend_from_slice(public_key.to_string().as_bytes());
```

Comment:
you can save an allocation by using [`PublicKey::as_str`](https://github.com/calimero-network/core/blob/24f0cc6bfea9476314bf57784a503cee05d1d0a5/crates/primitives/src/identity.rs#L123-L125)

```suggestion
                result.extend_from_slice(public_key.as_str().as_bytes());
```